### PR TITLE
[MLIR][Presburger] Implement computation of generating function for unimodular cones

### PR DESCRIPTION
**Pull request by Abhinav271828**
We implement a function that computes the generating function corresponding to a unimodular cone.
The generating function for a polytope is obtained by summing these generating functions over all tangent cones.

### CODE REVIEW

File: mlir/lib/Analysis/Presburger/Barvinok.cpp [hunk 1]
```diff
@@ -63,3 +63,71 @@ MPInt mlir::presburger::detail::getIndex(ConeV cone) {
 
   return cone.determinant();
 }
+
+/// Compute the generating function for a unimodular cone.
+GeneratingFunction mlir::presburger::detail::unimodularConeGeneratingFunction(
```
[the remark this answers is not in the document]
Comment (Abhinav271828) [reply mlir/lib/Analysis/Presburger/Barvinok.cpp:75]: Using unqualified names seems to lead to linker errors. I think definitions need to have them; simple references don't.

[the remark this answers is not in the document]
Comment (tobiasgrosser) [reply mlir/lib/Analysis/Presburger/Barvinok.cpp:75]: You would need to open the namespace ala:
```c++
namespace mlir {
   // here you can drop the qualifiers for definitions
}
```

[the remark this answers is not in the document]
Comment (Abhinav271828) [reply mlir/lib/Analysis/Presburger/Barvinok.cpp:75]: Right. I've been following what I thought was the convention in the library (e.g. https://github.com/llvm/llvm-project/blob/main/mlir/lib/Analysis/Presburger/IntegerRelation.cpp); is it not?

[the remark this answers is not in the document]
Comment (iambrj) [reply mlir/lib/Analysis/Presburger/Barvinok.cpp:75]: @Abhinav271828 just a nit, but notice how it is `IntegerRelation::setSpace` and not `mlir::presburger::IntegerRelation::setSpace`

[the remark this answers is not in the document]
Comment (iambrj) [reply mlir/lib/Analysis/Presburger/Barvinok.cpp:75]: also, my guess is the linker errors have something to do with makefile 

> Using unqualified names seems to lead to linker errors. I think definitions need to have them; simple references don't.

[the remark this answers is not in the document]
Comment (Abhinav271828) [reply mlir/lib/Analysis/Presburger/Barvinok.cpp:75]: oh that's true.. I'm not sure why I'm having linker errors then. Will look into this!

File: mlir/lib/Analysis/Presburger/Barvinok.cpp [hunk 2]
```diff
@@ -63,3 +63,78 @@ MPInt mlir::presburger::detail::getIndex(ConeV cone) {
...
+
+  unsigned numVar = cone.getNumVars();
+  unsigned numIneq = cone.getNumInequalities();
+
+  // Thus its ray matrix, U, is the inverse of the
+  // transpose of its inequality matrix, `cone`.
+  FracMatrix transp(numVar, numIneq);
+  for (unsigned i = 0; i < numVar; ++i)
+    for (unsigned j = 0; j < numIneq; ++j)
+      transp(j, i) = Fraction(cone.atIneq(i, j), 1);
```
Comment (Superty) [inline mlir/lib/Analysis/Presburger/Barvinok.cpp:88]: Can you add transpose as a function to matrix and use that?

File: mlir/lib/Analysis/Presburger/Barvinok.cpp [hunk 3]
```diff
@@ -63,3 +63,78 @@ MPInt mlir::presburger::detail::getIndex(ConeV cone) {
...
+GeneratingFunction mlir::presburger::detail::unimodularConeGeneratingFunction(
+    ParamPoint vertex, int sign, ConeH cone) {
+  // `cone` is assumed to be unimodular.
+  assert(getIndex(getDual(cone)) == 1 && "input cone is not unimodular!");
+
+  unsigned numVar = cone.getNumVars();
+  unsigned numIneq = cone.getNumInequalities();
+
+  // Thus its ray matrix, U, is the inverse of the
+  // transpose of its inequality matrix, `cone`.
```
Comment (Superty) [inline mlir/lib/Analysis/Presburger/Barvinok.cpp:84]: Inverse not computed?

File: mlir/lib/Analysis/Presburger/Barvinok.cpp [hunk 4]
```diff
@@ -63,3 +63,78 @@ MPInt mlir::presburger::detail::getIndex(ConeV cone) {
...
+  // are given by the generators of the cone, i.e.,
+  // the rows of the matrix U.
+  std::vector<Point> denominator(numIneq);
+  ArrayRef<Fraction> row;
+  for (unsigned i = 0; i < numVar; ++i) {
+    row = generators.getRow(i);
+    denominator[i] = Point(row);
+  }
+
+  // The vertex is v : [d, n+1].
```
Comment (Superty) [inline mlir/lib/Analysis/Presburger/Barvinok.cpp:103]: I can't understand this

File: mlir/lib/Analysis/Presburger/Barvinok.cpp [hunk 5]
```diff
@@ -63,3 +63,78 @@ MPInt mlir::presburger::detail::getIndex(ConeV cone) {
...
+  // We need to find affine functions of parameters λi(p)
```
Comment (Superty) [inline mlir/lib/Analysis/Presburger/Barvinok.cpp:104]: lambda_i?

File: mlir/lib/Analysis/Presburger/Barvinok.cpp [hunk 6]
```diff
@@ -63,3 +63,78 @@ MPInt mlir::presburger::detail::getIndex(ConeV cone) {
...
+  // such that v = Σ λi(p)*ui.
```
Comment (Superty) [inline mlir/lib/Analysis/Presburger/Barvinok.cpp:105]: u_i

File: mlir/lib/Analysis/Presburger/Barvinok.cpp [hunk 7]
```diff
@@ -63,3 +63,78 @@ MPInt mlir::presburger::detail::getIndex(ConeV cone) {
...
+  // The λi are given by the columns of Λ = v^T @ U^{-1} = v^T @ transp.
```
Comment (Superty) [inline mlir/lib/Analysis/Presburger/Barvinok.cpp:106]: I think concatenation is fine for matrix multiplication, @ is not used anywhere else in FPL

File: mlir/lib/Analysis/Presburger/Barvinok.cpp [hunk 8]
```diff
@@ -63,3 +63,78 @@ MPInt mlir::presburger::detail::getIndex(ConeV cone) {
...
+  // transpose of its inequality matrix, `cone`.
+  FracMatrix transp(numVar, numIneq);
+  for (unsigned i = 0; i < numVar; ++i)
+    for (unsigned j = 0; j < numIneq; ++j)
+      transp(j, i) = Fraction(cone.atIneq(i, j), 1);
+
+  FracMatrix generators(numVar, numIneq);
+  transp.determinant(&generators); // This is the U-matrix.
+
+  // The denominators of the generating function
```
Comment (Superty) [inline mlir/lib/Analysis/Presburger/Barvinok.cpp:93]: "Powers in the denominator" would be cleare

File: mlir/lib/Analysis/Presburger/Barvinok.cpp [hunk 9]
```diff
@@ -63,3 +63,78 @@ MPInt mlir::presburger::detail::getIndex(ConeV cone) {
...
+  for (unsigned i = 0; i < numVar; ++i) {
+    row = generators.getRow(i);
+    denominator[i] = Point(row);
+  }
+
+  // The vertex is v : [d, n+1].
+  // We need to find affine functions of parameters λi(p)
+  // such that v = Σ λi(p)*ui.
+  // The λi are given by the columns of Λ = v^T @ U^{-1} = v^T @ transp.
+  // Then the numerator will be Σ -floor(-λi(p))*u_i.
```
Comment (Superty) [inline mlir/lib/Analysis/Presburger/Barvinok.cpp:107]: Please make it clear you're talking about exponents

File: mlir/lib/Analysis/Presburger/Barvinok.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (Abhinav271828) [reply mlir/lib/Analysis/Presburger/Barvinok.cpp:75]: Hmm, @Superty and I discussed it and decided that a `Barvinok` namespace might be appropriate for all these functions and classes. I'll add that in a later patch.

[the remark this answers is not in the document]
Comment (joker-eph) [reply mlir/lib/Analysis/Presburger/Barvinok.cpp:75]: > You would need to open the namespace ala:
> 
> ```c++
> namespace mlir {
>    // here you can drop the qualifiers for definitions
> }
> ```

This isn't as safe as fully-qualifying the functions.
When are `using namespace mlir;` and qualifying the namespace on the function definition, you get a compile-time error if there isn't a matching declaration in the header.

File: mlir/lib/Analysis/Presburger/Barvinok.cpp (hunk 2 above)
Comment (Superty) [reply mlir/lib/Analysis/Presburger/Barvinok.cpp:88]: You can expose that too as a getInequalities

File: mlir/lib/Analysis/Presburger/Barvinok.cpp [hunk 10]
```diff
@@ -63,3 +63,79 @@ MPInt mlir::presburger::detail::getIndex(ConeV cone) {
...
+  // function are given by the generators of the cone,
+  // i.e., the rows of the matrix U.
+  std::vector<Point> denominator(numIneq);
+  ArrayRef<Fraction> row;
+  for (unsigned i = 0; i < numVar; ++i) {
+    row = generators.getRow(i);
+    denominator[i] = Point(row);
+  }
+
+  // The vertex is v : d x (n+1)
```
Comment (Superty) [inline mlir/lib/Analysis/Presburger/Barvinok.cpp:103]: Z^{d x (n + 1)}?

File: mlir/lib/Analysis/Presburger/Barvinok.cpp [hunk 11]
```diff
@@ -63,3 +63,79 @@ MPInt mlir::presburger::detail::getIndex(ConeV cone) {
...
+  // We need to find affine functions of parameters λ_i(p)
+  // such that v = Σ λ_i(p)*u_i.
+  // The λi are given by the columns of Λ = v^T U^{-1} = v^T transp.
```
Comment (Superty) [inline mlir/lib/Analysis/Presburger/Barvinok.cpp:106]: lambda_i

Comment (Superty) [reply mlir/lib/Analysis/Presburger/Barvinok.cpp:106]: `v^T transp.`?

Comment (Abhinav271828) [reply mlir/lib/Analysis/Presburger/Barvinok.cpp:107]: v^T multiplied by the matrix transp, is it not clear? How do I rewrite it?

File: mlir/include/mlir/Analysis/Presburger/Matrix.h [hunk 12]
```diff
@@ -181,6 +181,9 @@ class Matrix {
   /// `elems` must be equal to the number of columns.
   unsigned appendExtraRow(ArrayRef<T> elems);
 
+  // Transpose the matrix without modifying it.
+  Matrix<T> transpose();
```
Comment (Superty) [inline mlir/include/mlir/Analysis/Presburger/Matrix.h:185]: `/// Return the transpose of the matrix.`

Add const qualifier

File: mlir/lib/Analysis/Presburger/Barvinok.cpp (hunk 11 above)
Comment (Superty) [reply mlir/lib/Analysis/Presburger/Barvinok.cpp:107]: ok, I thought `transp.` was abbreviation of the word "transpose". It's easier to follow if the documentation names are all matched to the variable names, you could change either to match the other

Comment (Abhinav271828) [reply mlir/lib/Analysis/Presburger/Barvinok.cpp:107]: wait it does, the variable name is `transp`
should I name it something else?

Comment (Superty) [reply mlir/lib/Analysis/Presburger/Barvinok.cpp:108]: `The λi` -> `The λ_i`

Comment (Superty) [reply mlir/lib/Analysis/Presburger/Barvinok.cpp:108]: yeah I meant the other variables (matrix `U` in particular). But it's a bit hard to write these equations with long names

Comment (Superty) [reply mlir/lib/Analysis/Presburger/Barvinok.cpp:108]: I guess for now just write `Λ = v^T U^{-1}, and we have transp = U^{-1}` so it's clear

File: mlir/lib/Analysis/Presburger/Barvinok.cpp [hunk 13]
```diff
@@ -63,3 +64,80 @@ MPInt mlir::presburger::detail::getIndex(ConeV cone) {
...
+  std::vector<Point> denominator(numIneq);
+  ArrayRef<Fraction> row;
+  for (auto i : llvm::seq<int>(0, numVar)) {
+    row = generators.getRow(i);
+    denominator[i] = Point(row);
+  }
+
+  // The vertex is v \in Z^{d x (n+1)}
+  // We need to find affine functions of parameters λ_i(p)
+  // such that v = Σ λ_i(p)*u_i.
```
Comment (Superty) [inline mlir/lib/Analysis/Presburger/Barvinok.cpp:107]: what are u_i?

File: mlir/lib/Analysis/Presburger/Barvinok.cpp (hunk 11 above)
Comment (Abhinav271828) [reply mlir/lib/Analysis/Presburger/Barvinok.cpp:108]: okay

File: mlir/lib/Analysis/Presburger/Barvinok.cpp (hunk 13 above)
Comment (Abhinav271828) [reply mlir/lib/Analysis/Presburger/Barvinok.cpp:107]: generators, rows of U

Comment (Superty) [reply mlir/lib/Analysis/Presburger/Barvinok.cpp:107]: Cool, please mention it in the doc

File: mlir/lib/Analysis/Presburger/Barvinok.cpp [hunk 14]
```diff
@@ -63,3 +64,82 @@ MPInt mlir::presburger::detail::getIndex(ConeV cone) {
...
+  // and the cone has H-representation [0  -1] => U-matrix [ 2 -1]
+  //                                   [-1 -2]             [-1  0]
+  // Therefore Λ will be given by [ 1    0 ] and the negation of this will be
+  // stored as the numerator.
+  //                              [ 1/2 -1 ]
+  //                              [ -1  -2 ]
+
+  // Algebraically, the numerator exponent is
+  // [ -2 ⌊ -N - M/2 +1 ⌋ + 1 ⌊ 0 +M +2 ⌋ ] -> first  COLUMN of U is [2, -1]
+  // [  1 ⌊ -N - M/2 +1 ⌋ + 0 ⌊ 0 +M +2 ⌋ ] -> second COLUMN of U is [-1, 0]
```
Comment (Superty) [inline mlir/lib/Analysis/Presburger/Barvinok.cpp:144]: I cannot understand this representation on the left side if they're supposed to be expressions then don't write it as `0 +M +2`, it looks like a matrix?

File: mlir/lib/Analysis/Presburger/Barvinok.cpp [hunk 15]
```diff
@@ -63,3 +64,84 @@ MPInt mlir::presburger::detail::getIndex(ConeV cone) {
...
+  // i.e., its generators are given by U = [2  -1].
+  //                                       [-1  0]
+  // Therefore Λ will be given by [ 1    0 ] and the negation of this will be
+  // stored as the numerator.
+  //                              [ 1/2 -1 ]
+  //                              [ -1  -2 ]
+
+  // Algebraically, the numerator exponent is
+  // [ -2 ⌊ - N - M/2 + 1 ⌋ + 1 ⌊ 0 + M + 2 ⌋ ] -> first  COLUMN of U is [2, -1]
+  // [  1 ⌊ - N - M/2 + 1 ⌋ + 0 ⌊ 0 + M + 2 ⌋ ] -> second COLUMN of U is [-1, 0]
```
Comment (Superty) [inline mlir/lib/Analysis/Presburger/Barvinok.cpp:146]: This would probably by easier to follow if it was in-line with the rest of the explanation

**Review by superty — APPROVED** [review]

**Pull request merged**